plan/logical_plans.go

```diff
@@ -359,32 +401,110 @@ func (ds *DataSource) deriveIndexPathStats(path *accessPath) error {
 	sc := ds.ctx.GetSessionVars().StmtCtx
 	path.ranges = ranger.FullRange()
 	path.countAfterAccess = float64(ds.statisticTable.Count)
-	idxCols, lengths := expression.IndexInfo2Cols(ds.schema.Columns, path.index)
-	if len(idxCols) != 0 {
-		path.ranges, path.accessConds, path.indexFilters, path.eqCondCount, err = ranger.DetachCondAndBuildRangeForIndex(ds.ctx, ds.pushedDownConds, idxCols, lengths)
+	path.idxCols, path.colLengths = expression.IndexInfo2Cols(ds.schema.Columns, path.index)
+	if len(path.idxCols) != 0 {
+		path.ranges, path.accessConds, path.tableFilters, path.eqCondCount, err = ranger.DetachCondAndBuildRangeForIndex(ds.ctx, ds.pushedDownConds, path.idxCols, path.colLengths)
 		if err != nil {
 			return errors.Trace(err)
 		}
 		path.countAfterAccess, err = ds.statisticTable.GetRowCountByIndexRanges(sc, path.index.ID, path.ranges)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		path.indexFilters, path.tableFilters = splitIndexFilterConditions(path.indexFilters, path.index.Columns, ds.tableInfo)
 	} else {
-		path.indexFilters, path.tableFilters = splitIndexFilterConditions(ds.pushedDownConds, path.index.Columns, ds.tableInfo)
+		path.tableFilters = ds.pushedDownConds
+	}
+	corColInAccessConds := false
+	if path.eqCondCount == len(path.accessConds) {
+		access, remained := splitCorColAccessCondFromFilters(path.idxCols, path.colLengths, path.eqCondCount, path.tableFilters)
+		path.accessConds = append(path.accessConds, access...)
+		path.tableFilters = remained
+		if len(access) > 0 {
+			corColInAccessConds = true
+		}
+	}
+	path.indexFilters, path.tableFilters = splitIndexFilterConditions(path.tableFilters, path.index.Columns, ds.tableInfo)
+	if corColInAccessConds {
+		idxHist, ok := ds.statisticTable.Indices[path.index.ID]
+		if ok {
+			path.countAfterAccess = idxHist.AvgCountPerValue(ds.statisticTable.Count)
+		} else {
+			path.countAfterAccess = ds.statisticTable.PseudoAvgCountPerCount()
+		}
 	}
-	path.countAfterIndex = path.countAfterAccess
 	if path.indexFilters != nil {
 		selectivity, err := ds.statisticTable.Selectivity(ds.ctx, path.indexFilters)
 		if err != nil {
 			log.Warnf("An error happened: %v, we have to use the default selectivity", err.Error())
 			selectivity = selectionFactor
 		}
-		path.countAfterIndex = math.Max(path.countAfterAccess*selectivity, ds.statsAfterSelect.count)
+		path.countAfterIndex = math.Min(path.countAfterAccess*selectivity, ds.statsAfterSelect.count)
```

Comment:
This would make the table filter conditions look like do not filter anything, like the explain test indicates.

Reply:
ok, i'll check this.

Reply:
@lamxTyler There comes a question. `Selectivity` doesn't specify the correlated column. So when we calculate `ds.statsAfterSelect.count` the correlated part will be just treated as `selectionFactor` which is 0.8. So this will make `countAfterAccess` smaller than `countAfterIndex`, which is opposed to fact.
One solution is specify the correlated column in `Selectivity`, what do think?

Reply:
What about we first scale the `statsAfterSelect`? The selectivity of index filters and table filters is `countAfterSelect/countAfterAccess`.

Reply:
Emmm... I didn't get the point...

Reply:
I mean if the `countAfterAccess` is changed, then the `countAfterSelect` should be changed accordingly.
So, when we change the `countAfterAccess`, we should immediately scale the `statsAfterSelect`.